feat(berajs): add gas limit estimation to prevent hitting gas limits …

BrownBrownBear opened the pull request:
…set by validator nodes

bearpong commented on packages/berajs/src/hooks/useContractWrite/useBeraContractWrite.ts:
```diff
@@ -60,18 +62,31 @@ const useBeraContractWrite = ({
-            account: account,
-          });
+          const [{ request }, wagmiPubEstimateContractGas] = await Promise.all([
+            publicClient.simulateContract({
+              address: address,
+              abi: abi,
+              functionName: functionName,
+              args: params,
+              value: value,
+              account: account,
+            }),
+            publicClient.estimateContractGas({
+              address: address,
+              abi: abi,
+              functionName: functionName,
+              args: params,
+              value: value,
+              account: account,
+            }),
+          ]);
+
           receipt = await writeContractAsync({
             ...request,
-            gas: gasLimit ?? request.gas,
+            gas:
+              increaseByPercentage(wagmiPubEstimateContractGas, 10) ?? gasLimit,
```
Here I think we should:
1. remove default gasLimit value
2. don't fetch estimateContractGas unless gasLimit is unset
3. use Promise.allSettled and:
- throw error if simulateContract reverts
- use  DEFAULT_METAMASK_GAS_LIMIT if no gasLimit was provided and estimateContractGas failed

`increaseByPercentage` should be removed since we realized that gas is different from the gas cost. Attaching here a screenshot that confirms our hints 
<img width="654" alt="Screenshot 2024-12-16 at 11 40 41" src="https://github.com/user-attachments/assets/3406aa70-a38e-4071-afc6-9840043420f6" />

BrownBrownBear replied:
<img width="1457" alt="Screenshot 2024-12-19 at 11 01 51 AM" src="https://github.com/user-attachments/assets/8b866a47-31d9-4f32-9b15-dd46b676e609" />
For example, if I didn't bump this, you can see it's greater than 90% and the txn would not have gone through.

bearpong commented on packages/berajs/src/utils/const.ts:
```diff
@@ -0,0 +1 @@
+export const DEFAULT_METAMASK_GAS_LIMIT = 7920027n; // NOTE: this is the metamask gas limit, https://github.com/MetaMask/metamask-extension/blob/06cf7459b963e26d0ae60312d4f9a342d07ab89d/ui/pages/confirmations/send/send.constants.js#L8
```
can we use JSdocs for this?

bearpong reviewed (changes requested):
Please after addressing the comments, might you also add the nonce to the tx in this PR? So that we can test all changes to useBeraContractWrite at once.

It'd also be useful to check if useCreatePool works without the gasLimit hardcoded

BrownBrownBear commented:
<img width="1377" alt="Screenshot 2024-12-19 at 11 04 16 AM" src="https://github.com/user-attachments/assets/458fbe5b-6f33-4331-890e-0965546e34d9" />
I did attach my concerns with adding out own nonce but things work. And if people are trying to queue many txns, we can still calculate the pending txns!

BrownBrownBear commented:
> Please after addressing the comments, might you also add the nonce to the tx in this PR? So that we can test all changes to useBeraContractWrite at once.
> 
> It'd also be useful to check if useCreatePool works without the gasLimit hardcoded

I did check and simulate if useCreatePool does work. I just didn't create a pool since it can throw off the prices we have on boonet we have right now

bearpong reviewed (approved):
LGTM
